[Platforms] Add paragraph on TVP and "internal" (#308)
cncf/tag-app-delivery: platforms-whitepaper/v1alpha1/paper.md

platform teams minimize resources and cognitive load of the stream-aligned
[product] team... platform teams can create a cohesive experience that spans
across different user experiences or products."

According to Martin Fowler and Evan Bottcher [[2]], "a digital platform is a 
foundation of self-service APIs, tools, services, knowledge and support which 
are arranged as a compelling internal product. Autonomous delivery teams can 
make use of the platform to deliver product features at a higher pace, with 
reduced coordination."

The specific set of capabilities provided by a platform depends on the
requirements of the platform's users. And while platforms _provide_ these
required capabilities, it's critical to note that platform teams should not
always _implement_ them themselves. Managed service providers or dedicated
internal teams can maintain backing implementations while platforms are the
thinnest reasonable layer that provides consistency for these implementations
that meets an organization's requirements. For example, a very simple "platform"
could be a wiki page with links to standard operating procedures to provision
capabilities from providers, as described in [[3]].

> Because these platforms target no more and no less than an enterprise's internal

Author: I think best not to mention it since we don't intend to _exclude_ it - we just believe it's _one_ of many types of platforms.

make use of the platform to deliver product features at a higher pace, with 
reduced coordination."

The specific set of capabilities provided by a platform depends on the
requirements of the platform's users. And while platforms _provide_ these
required capabilities, it's critical to note that the platform team generally

> shouldn't _implement_ them themselves. Managed providers or dedicated teams

Reviewer: I agree here as well - whether they choose to implement or not depends on their business needs

> shouldn't _implement_ them themselves. Managed providers or dedicated teams

Author: Fixed.

should maintain implementations while the platform itself should be the thinnest
consistent layer that meets an organization's requirements. For example, the
simplest "platform" would be a wiki page with links to provision capabilities
from providers. Some [[3]] have described this principle as seeking the

> "thinnest viable platform".

Reviewer: again saying here that a platform should always aspire to stay thin isn't necessarily correct here 

in the Gartner paper for example when it talks about scaling platform teams a platform should START as thin but can evolve - suggesting it stays as thin as possible doesn't send the right message because a platform can likely evolve over time based on the business needs

> "thinnest viable platform".

Reviewer 2: Thanks for adding this Abhinav, this is the difference in definition I was hoping to explore!

I would be uncomfortable using a new term to describe a long staging idea of starting small. Product may be new to internal platforms, but it is not new and we can add should lean on the existing work where possible. 

Therefore, I'd advocate that if we keep the wording in the space if "start small and then grow big" that we focus on that being defined at ["Minimum Viable Product"](https://en.wikipedia.org/wiki/Minimum_viable_product) on first delivery. 

I would personally prefer to see us use the term "Thinnest Viable Platform" but do so with a focus on how [Gregor Hope described the functionally as floating](https://m.youtube.com/watch?v=WaL3ZbLgMuI). Always staying as thin as possible to provide the required functionally.

> "thinnest viable platform".

Reviewer 3: Also, I see MVP and TVP as useful principles that can and perhaps should coexist here. The ideas behind MVP shouldn't be new since starting small and working in and iterative way is what we usually do in software engineering. Perhaps the new thing is applying those ideas consistently to an internal platform product. 
Instead, TVP is something new that can help drive architectural decisions for a platform and avoid the trap of reinventing the wheel or starting all from scratch instead of embracing what the community and vendors have to offer out there.

> "thinnest viable platform".

Author: I adjusted the language a bit here to avoid referring to TVP explicitly since the term may have/develop it's own connotations. I added some language in the Platform Teams section to further reflect the idea of minimizing the platform layer, PTAL.

I love the idea of the "floating" platform and continually delegating more to managed service providers, though we don't have to mention it explicitly at this point. Thinking to leave this proposal as is for now and we can consider expanding it later.

Because these platforms target no more and no less than an enterprise's internal

> users we often refer to them as _internal_ platforms.

Reviewer 2: 💜 Love this. Super simple, but clarifying.

Reviewer: agreed I think I can integrate this into what I have

Reviewer 3: I really like this addition. Very effective!

Author: Awesome, thanks for the confirmation!

   spaces and supporting services to deploy and use those third-party products and 
   services

A more advanced internal platform also _composes_ capabilities into experiences
and templates fit for key scenarios like web application development or data

> analysis. For example, an internal developer platform includes templates for
> provisioning complete development environments and could serve the following use
> cases:

Reviewer 2: I just clocked that we talk about templates here and not on-demand services. IMO if we are describing "mature" here we need to step away from templates (which still leave ownership with the consuming team to use them) and all the way to on demand.

```suggestion
analysis. For example, an internal developer platform provide on demand 
provisioning of complete development environments and could serve the following use
cases:
```

Author: I hear you and will try to add self-service and other short adjectives here. I think I'm assuming that templates, like APIs, depend on a self-serviceable implementation that reifies whatever they declare, so perhaps I can express that.

Author: In my last commit I combined the two lists of use cases and mentioned self-service and templates, PTAL!

Following are jobs a platform team should be responsible for:

1. Research platform user requirements and plan feature roadmap
1. Market, evangelize and advocate for the platform's proposed values
1. Manage and develop interfaces for using and observing capabilities and
   services, including portals, APIs, documentation and templates, and CLI tools

Most importantly, platform teams must learn about the requirements of platform
users to inform and continuously improve capabilities and interfaces offered by
their platform. Ways to learn about user requirements include user interviews,
interactive hackathons, issue trackers and surveys, and direct observation of
usage through observability tools. For example, a platform team could publish a

> form for users to submit feature requests; and conduct periodic roadmap meetings

Author: :+1: I hear you, will remove the semicolon and add those suggestions.

> form for users to submit feature requests; and conduct periodic roadmap meetings

Author: Removed semicolon and added observe metrics to list :+1:

activities including broad announcements, engaging demos, and regular feedback
and communication sessions.  The key here is to meet users where they are and
bring them on a journey to engage with and benefit from the platform.

A platform team doesn't necessarily run compute, network, storage or other
services. In fact an internal platform should rely on _externally_-provided
services and capabilities as much as possible; platform teams should build and
maintain their own capabilities only when they're not available elsewhere from
managed providers or internal infrastructure teams. Instead, platform teams are
most responsible for the _interfaces_ (i.e., GUI, CLI, and API) and user
experiences for the services and capabilities their platform makes available.

For example, a Web page in a platform might describe and even offer a button to
provision an identity for an app; while the implementation of that capability
might be via a cloud-hosted identity service. An internal platform team may
manage the web page and an API, but not the actual service implementation.
Platform teams should consider creating and maintaining their own capabilities

> only when a required capability is not available elsewhere.

Author: That's a good point, I'll think of a way to include that.

> only when a required capability is not available elsewhere.

Author: Added the word "generally consider..." so that this statement isn't absolute.

roadmap meetings to share upcoming features and gather feedback.

A platform team doesn't necessarily run compute, network, storage or other

> services, and in fact an internal platform should be as thin as viable as

Reviewer: not sure I agree with this 
an internal platform should START as thin as viable but can expand based on the needs of an organization

> services, and in fact an internal platform should be as thin as viable as
> described above. A platform team instead is most responsible for the
> _interfaces_ \[2\] (GUI, CLI, API) and user experiences with those services. For
> example, a Web portal page might describe and even offer a button to provision
> an identity for an app; while the implementation of that capability might be via
> a cloud-hosted identity service.  An internal platform team may manage the web
> page and user-facing API, but not necessarily the implementation.

Reviewer 2: ```suggestion
services, and in fact an internal platform should maintain a minimal footprint as
described above. A platform team instead is most responsible for the
_interfaces_ \[2\] (GUI, CLI, API) and user experiences with those services. For
example, a Web portal page might describe and even offer a button to provision
an identity for an app; while the implementation of that capability might be via
a cloud-hosted identity service.  An internal platform team may manage the web
page and user-facing API, but not necessarily the implementation. Only when a 
required capability is not available in the market should a platform team create 
and manage its complexity in house.
```

Author: Took a shot at updating in latest commit, PTAL!
